SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/PredictedGameplayCuesTest/PredictedGameplayCuesTest.cpp

```diff
@@ -0,0 +1,171 @@
+#include "GameplayCueSet.h"
+#include "Kismet/GameplayStatics.h"
+#include "SpatialFunctionalTestFlowController.h"
+
+/**
+ * Tests that gameplay cue events correctly trigger on all clients when triggered by a predicted gameplay effect application.
+ * - Setup:
+ *	- One server, two clients.
+ *	- The server spawns the test pawn, which has an Ability System Component (ASC) and the ability to be activated already granted to it.
+ *	- The test pawn is possessed by client 1's player controller. This allows client 1 to activate predicted abilities on the pawn.
+ *	- All workers wait until the test pawn has been replicated to them.
+ * - Test:
+ *	- Client 1 activates the ability on the pawn through a gameplay event with the "execute" tag.
+ *	  Based on the tag, the activated ability applies an instant gameplay effect, which has the gameplay cue to be tested assigned to it.
+ *	  This should trigger an "execute" event on the gameplay cue.
+ *  - All clients wait until they have seen then gameplay cue receive an "execute" event exactly once.
+ *  - Client 1 activates the ability with the "activate" tag.
+ *	  When activated with this tag, the ability applies a gameplay effect with duration,
+ *    which should cause "OnActive", "WhileActive" and "Remove" events on the gameplay cue.
+ *	- All clients wait until they have seen the gameplay cue receive an "OnActive" event exactly once.
+ * - Cleanup:
+ *	- The original pawn is re-possessed by client 1's player controller.
+ *	- The test pawn is registered to be auto-destroyed by the test framework at test end.
+ */
+APredictedGameplayCuesTest::APredictedGameplayCuesTest()
+{
+	Author = TEXT("Tilman Schmidt");
+	Description =
+		TEXT("Tests that gameplay cue events correctly trigger on all clients when triggered by a predicted gameplay effect application.");
+	DuplicateActivationCheckWaitTime = 2.0f;
+
+	TestPawn = nullptr;
+	StepTimer = -1.0f;
+}
+
+void APredictedGameplayCuesTest::PrepareTest()
+{
+	Super::PrepareTest();
+
+	AddStep(TEXT("Register gameplay cue"), FWorkerDefinition::AllClients, nullptr, [this]() {
+		// A code-only gameplay cue will not get automatically picked up by the gameplay cue manager
+		// (by default it searches the game content folder for assets).
+		// If a cue is not registered with the manager, cue events will never be routed to it.
+		// That means we have to register the cue manually here on each client.
+		FSoftObjectPath CuePath(TEXT("/Script/SpatialGDKFunctionalTests.GC_SignalCueActivation"));
+		FGameplayCueReferencePair ExecuteCueRef(UGC_SignalCueActivation::GetExecuteTag(), CuePath);
+		FGameplayCueReferencePair AddCueRef(UGC_SignalCueActivation::GetAddTag(), CuePath);
+		UAbilitySystemGlobals::Get().GetGameplayCueManager()->GetRuntimeCueSet()->AddCues({ ExecuteCueRef, AddCueRef });
+
+		FinishStep();
+	});
+
+	auto WaitForEventConfirmation = [this](int Counter, float DeltaTime) {
+		// StepTimer == -1.0f signals that we have not yet seen Counter be 1
```

Comment:
if a sentinel value is unavoidable I would define named static float called SENTINEL_VALUE or similar.

Reply:
Changed it to be two variables instead now, for clarity.

Reply:
I think we miss understood each other, what I meant is something like:
```suggestion
		static float TIMER_NOT_STARTED = -1.0f;
		if (StepTimer == TIMER_NOT_STARTED )
```

Reply:
I got what you meant by it, but I felt that a separate variable might be clearer? Downside is that now they could be out of sync. Not really sure which way to make the call here.

Reply:
Oh wait, I messed something up and the changes got lost 😬 Will re-do, then what I'm saying should make sense.

Reply:
Yeah I saw those later changes now, makes sense to me.